server/src/main/java/org/elasticsearch/snapshots/SnapshotDeletionsPendingExecutor.java

```diff
@@ -0,0 +1,451 @@
+import org.elasticsearch.cluster.RestoreInProgress;
+import org.elasticsearch.cluster.SnapshotDeletionsPending;
+import org.elasticsearch.cluster.SnapshotsInProgress;
+import org.elasticsearch.cluster.metadata.RepositoriesMetadata;
+import org.elasticsearch.cluster.metadata.RepositoryMetadata;
+import org.elasticsearch.cluster.service.ClusterApplierService;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.AbstractRunnable;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+import org.elasticsearch.common.util.concurrent.CountDown;
+import org.elasticsearch.common.util.set.Sets;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.repositories.RepositoryData;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.time.Instant;
+import java.time.ZoneOffset;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.repositories.blobstore.BlobStoreRepository.READONLY_SETTING_KEY;
+import static org.elasticsearch.snapshots.SnapshotUtils.cloneSources;
+import static org.elasticsearch.snapshots.SnapshotUtils.deletionsSources;
+import static org.elasticsearch.snapshots.SnapshotUtils.restoreSources;
+import static org.elasticsearch.snapshots.SnapshotsService.findRepositoryForPendingDeletion;
+
+public class SnapshotDeletionsPendingExecutor {
+
+    private static final Logger logger = LogManager.getLogger(SnapshotDeletionsPendingExecutor.class);
+
+    public static final Setting<TimeValue> PENDING_SNAPSHOT_DELETIONS_RETRY_INTERVAL_SETTING = Setting.timeSetting(
+        "snapshot.snapshot_deletions_pending.retry_interval",
+        TimeValue.timeValueSeconds(30L),
+        TimeValue.ZERO,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+
+    public static final Setting<TimeValue> PENDING_SNAPSHOT_DELETIONS_EXPIRATION_INTERVAL_SETTING = Setting.timeSetting(
+        "snapshot.snapshot_deletions_pending.expiration_interval",
+        TimeValue.timeValueHours(12L),
+        TimeValue.ZERO,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+
+    /**
+     * Set of pending snapshots deletions whose deletion is already triggered
+     */
+    private final Set<SnapshotId> triggered = ConcurrentCollections.newConcurrentSet();
+
+    /**
+     * Map of pending snapshots deletions whose deletion is conflicting with on-going restores/clones/repository clean up or repository
+     * missing or read-only. Those sets are used to identify the cluster state updates to process in case they resolve some conflict.
+     */
+    private final Map<SnapshotId, ConflictType> conflicting = new HashMap<>();
```

Comment:
This replaces the different sets for the various type of conflicts that were introduced in this PR.